### ISSUE TITLE
Running motion correction on large dataset gets stuck with memory issues

### ISSUE THREAD

**Issue by Cumol** [flatironinstitute/CaImAn#201]
Ran the motion correction demo notebook without a problem. When I changed the path to one of my .tif files (5Gb in size) the script hangs and nothing happened for 3 hours. The machine has 36 cores and 50Gb of RAM, Python 3.6 on Ubuntu. This is the last message I get (the script does not stop):

```
** Startting parallel motion correction **
Exception in thread Thread-12:
Traceback (most recent call last):
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/multiprocessing/pool.py", line 405, in _handle_workers
    pool._maintain_pool()
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/multiprocessing/pool.py", line 246, in _maintain_pool
    self._repopulate_pool()
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/multiprocessing/pool.py", line 239, in _repopulate_pool
    w.start()
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/multiprocessing/context.py", line 277, in _Popen
    return Popen(process_obj)
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/multiprocessing/popen_fork.py", line 26, in __init__
    self._launch(process_obj)
  File "/home/generalubuntu/anaconda3/envs/caiman/lib/python3.6/multiprocessing/popen_fork.py", line 73, in _launch
    self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory
```

When I stop the script I get the following output:

```
---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-20-d74c81ca51c3> in <module>()
      1 # correct for rigid motion correction and save the file (in memory mapped form)
----> 2 mc.motion_correct_rigid(save_movie=True)
      3 # load motion corrected movie
      4 m_rig = cm.load(mc.fname_tot_rig)
      5 bord_px_rig = np.ceil(np.max(mc.shifts_rig)).astype(np.int)

~/CaImAn/caiman/motion_correction.py in motion_correct_rigid(self, template, save_movie)
    216                 add_to_movie=-self.min_mov,
    217                 nonneg_movie=self.nonneg_movie,
--> 218                 gSig_filt=self.gSig_filt)
    219 
    220             self.total_template_rig = _total_template_rig

~/CaImAn/caiman/motion_correction.py in motion_correct_batch_rigid(fname, max_shifts, dview, splits, num_splits_to_process, num_iter, template, shifts_opencv, save_movie_rigid, add_to_movie, nonneg_movie, gSig_filt, subidx)
   2390                                                              dview=dview, save_movie=save_movie, base_name=os.path.split(
   2391                                                                  fname)[-1][:-4] + '_rig_', subidx = subidx,
-> 2392                                                              num_splits=num_splits_to_process, shifts_opencv=shifts_opencv, nonneg_movie=nonneg_movie, gSig_filt=gSig_filt)
   2393 
   2394         new_templ = np.nanmedian(np.dstack([r[-1] for r in res_rig]), -1)

~/CaImAn/caiman/motion_correction.py in motion_correction_piecewise(fname, splits, strides, overlaps, add_to_movie, template, max_shifts, max_deviation_rigid, newoverlaps, newstrides, upsample_factor_grid, order, dview, save_movie, base_name, subidx, num_splits, shifts_opencv, nonneg_movie, gSig_filt)
   2670         print('** Startting parallel motion correction **')
   2671         if 'multiprocessing' in str(type(dview)):
-> 2672             res = dview.map_async(tile_and_correct_wrapper, pars).get(4294967)
   2673         else:
   2674             res = dview.map_sync(tile_and_correct_wrapper, pars)

~/anaconda3/envs/caiman/lib/python3.6/multiprocessing/pool.py in get(self, timeout)
    636 
    637     def get(self, timeout=None):
--> 638         self.wait(timeout)
    639         if not self.ready():
    640             raise TimeoutError

~/anaconda3/envs/caiman/lib/python3.6/multiprocessing/pool.py in wait(self, timeout)
    633 
    634     def wait(self, timeout=None):
--> 635         self._event.wait(timeout)
    636 
    637     def get(self, timeout=None):

~/anaconda3/envs/caiman/lib/python3.6/threading.py in wait(self, timeout)
    549             signaled = self._flag
    550             if not signaled:
--> 551                 signaled = self._cond.wait(timeout)
    552             return signaled
    553 

~/anaconda3/envs/caiman/lib/python3.6/threading.py in wait(self, timeout)
    297             else:
    298                 if timeout > 0:
--> 299                     gotit = waiter.acquire(True, timeout)
    300                 else:
    301                     gotit = waiter.acquire(False)
```

**Comment by Cumol** [comment]
@pgunn 
Checking for 32 vs. 64.

```
In [1]: import platform
   ...: platform.architecture()
   ...: 
Out[1]: ('64bit', '')
```

And running `ulimit -a` has the following output:

```
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 78139
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 78139
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

**Comment by pgunn** [comment]
Darn. Was worth a shot. One thing that's contributing to the memory usage is the worker model you're using; I believe under the hood it's doing a real fork() rather than threads to manage its workload, and this is much more memory-expensive than the other options (every process gets its own distinct copy of in-memory structures). You may want to investigate other options.

**Comment by epnev** [comment]
@Cumol  There is a potential similar issue although it might be just mac related: https://github.com/flatironinstitute/CaImAn/issues/206
Can you put the following commands in the first cell of your notebook *after* you import caiman:
```
import multiprocessing as mp
mp.set_start_method('forkserver', force=True)
```
Does this work?

**Comment by epnev** [comment]
@Cumol Another possible issue is that there is weird stuff happening due to multithreaded operations occurring in parallel. You can turn multithreading off by running:
```
export MKL_NUM_THREADS=1
export OPENBLAS_NUM_THREADS=1
```
on a terminal (or include it in your bash profile file)

**Comment by Cumol** [comment]
Thank you for your suggestions @epnev !

I tried the following:
```
import multiprocessing as mp
mp.set_start_method('forkserver', force=True)

```

And got 30 iterations of:

```
from ._conv import register_converters as _register_converters
```

Which then ended with `killed` with a full SWAP file on a 30Gb RAM ubuntu machine.

The second setting in the .bashrc file were already set as I followed the setup according to the readme. When I set them both to 0 I got a pipe error (?)

Any other suggestions?

I see you updated CaImAn in the meantime, should I update mine too?
